adaptdl/adaptdl/checkpoint.py

```diff
@@ -95,24 +107,45 @@ def save_all_states():
     for state in _STATES_TO_NAMES:
         save_state(state)
 
+    # Prevent corrupting original state files in case the process got killed
+    # during state file writing.
+    if replica_rank() == 0 and checkpoint_path() is not None:
+        tmp_ckpt_dir = _get_tmp_ckpt_dir()
+        ckpt_dir = os.path.join(checkpoint_path(),
+                                f"{CKPT_DIR_PREFIX}{num_restarts()}")
+        os.makedirs(ckpt_dir, exist_ok=True)
+        os.rename(tmp_ckpt_dir, ckpt_dir)  # atomic
+
+        for old_ckpt_dir in os.listdir(checkpoint_path()):
+            if (
+                os.path.isdir(old_ckpt_dir) and
+                old_ckpt_dir.startswith(CKPT_DIR_PREFIX) and
+                old_ckpt_dir != ckpt_dir
+            ):
+                shutil.rmtree(old_ckpt_dir)
```

Comment:
`os.listdir` will output just the directory name without the path, so you'll need to prepend the path to it.

Reply:
Thanks. Updated